File: VB/XAF.LinqToXpo/LinqCollectionSource.vb
Imports System
Imports System.Linq
Imports DevExpress.Xpo
Imports System.Collections
Imports System.ComponentModel
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.DC
Imports DevExpress.Data.Filtering
Imports DevExpress.ExpressApp.Xpo

Namespace XAF.LinqToXpo
    Public Class LinqCollectionSource
        Inherits CollectionSourceBase

        Public Const DefaultSuffix As String = "_Linq"
        Private objectTypeInfoCore As ITypeInfo
        Private collectionCore As BindingList(Of Object)
        Protected Sub New(ByVal objectSpace As IObjectSpace, ByVal mode As CollectionSourceMode)
            MyBase.New(objectSpace, mode)
        End Sub
        Protected Sub New(ByVal objectSpace As IObjectSpace)
            MyBase.New(objectSpace)
        End Sub
        Public Sub New(ByVal objectSpace As IObjectSpace, ByVal objectType As Type, ByVal query As IQueryable)
            MyBase.New(objectSpace)
            objectTypeInfoCore = XafTypesInfo.Instance.FindTypeInfo(objectType)
            Me.Query = query
        End Sub
        Private privateQuery As IQueryable
        Public Property Query() As IQueryable
            Get
                Return privateQuery
            End Get
            Private Set(ByVal value As IQueryable)
                privateQuery = value
            End Set
        End Property
        Protected Overrides Function CreateCollection() As Object
            DirectCast(Query, XPQueryBase).Session = CType(ObjectSpace, XPObjectSpace).Session
            Return ConvertQueryToCollection(Query)
        End Function
        Public Function ConvertQueryToCollection(ByVal sourceQuery As IQueryable) As IList
            collectionCore = New BindingList(Of Object)()
            For Each item In sourceQuery
                collectionCore.Add(item)
            Next item
            Return collectionCore
        End Function
        Public Overrides Function IsObjectFitForCollection(ByVal obj As Object) As Boolean?
            Return collectionCore.Contains(obj)
        End Function
        Public Overrides ReadOnly Property ObjectTypeInfo() As ITypeInfo
            Get
                Return objectTypeInfoCore
            End Get
        End Property
        Public Overrides Sub Dispose()
            MyBase.Dispose()
            collectionCore = Nothing
            Query = Nothing
        End Sub
        Protected Overrides Sub ApplyCriteriaCore(ByVal criteria As CriteriaOperator)
            ' This method has no implementation.
        End Sub
    End Class
End Namespace

File: CS/XAF.LinqToXpo/LinqCollectionSource.cs
using System;
using System.Linq;
using DevExpress.Xpo;
using System.Collections;
using System.ComponentModel;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.DC;
using DevExpress.Data.Filtering;
using DevExpress.ExpressApp.Xpo;

namespace XAF.LinqToXpo {
    public class LinqCollectionSource : CollectionSourceBase {
        public const string DefaultSuffix = "_Linq";
        private ITypeInfo objectTypeInfoCore;
        private BindingList<object> collectionCore;
        protected LinqCollectionSource(IObjectSpace objectSpace, CollectionSourceMode mode)
            : base(objectSpace, mode) {
        }
        protected LinqCollectionSource(IObjectSpace objectSpace)
            : base(objectSpace) {
        }
        public LinqCollectionSource(IObjectSpace objectSpace, Type objectType, IQueryable query)
            : base(objectSpace) {
            objectTypeInfoCore = XafTypesInfo.Instance.FindTypeInfo(objectType);
            Query = query;
        }
        public IQueryable Query {
            get;
            private set;
        }
        protected override object CreateCollection() {
            ((XPQueryBase)Query).Session = ((XPObjectSpace)ObjectSpace).Session;
            return ConvertQueryToCollection(Query);
        }
        public IList ConvertQueryToCollection(IQueryable sourceQuery) {
            collectionCore = new BindingList<object>();
            foreach (var item in sourceQuery) { collectionCore.Add(item); }
            return collectionCore;
        }
        public override bool? IsObjectFitForCollection(object obj) {
            return collectionCore.Contains(obj);
        }
        public override ITypeInfo ObjectTypeInfo {
            get { return objectTypeInfoCore; }
        }
        public override void Dispose() {
            base.Dispose();
            collectionCore = null;
            Query = null;
        }
        protected override void ApplyCriteriaCore(CriteriaOperator criteria) { 
            // This method has no implementation.
        }
    }
}